Enforce stricter competition names - fixes #444

pingskills opened the pull request:
Only allow alphanumeric characters, dash, ampersand, period,
colon, apostrophe, and space for name and cellName.

Update old invalid competition names so they are valid.

jfly commented:
Thanks for writing this migration! I like that we'll have this in our history for forever :)

Do we need to update the hint text for this field as well?

Can you reach out to the board and ask them how the feel about deploying this change? Feel free to cc software.

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -2,11 +2,24 @@
+    competition = FactoryGirl.build :competition, name: "Campeonato de Cubos Mágicos de São Carlos/SP 2013"
+    expect(competition).to be_invalid
+    competition = FactoryGirl.build :competition, name: "Moldavian Nationals – Winter 2016"
+    expect(competition).to be_invalid
```
I think this can be really cleaned up by a loop
```ruby
%w(name1, name2, ...).each do |name|
  expect(FactoryGirl.build(:competition, name: name).to be_invalid
end
```

pingskills commented:
I cleaned up the tests and added some hint text. What are your thoughts on the hint text? Does it make the hint too long? I also added the same error message for the cellName.

Travis is failing with an error message about wca_test not existing? And it also mentions that I have an outstanding migration. Is there something special I need to do for a pull request where there is a migration? Or has something changed and we really don't have a wca_test database for Travis?

jonatanklosko commented:
You have to run `rake db:migrate`.

pingskills commented:
@jonatanklosko I'm not really sure how to do that for Travis-ci? Is there any documentation you could point me at?

jonatanklosko commented:
Oh, I meant that you should do this locally, this will update structure.sql automatically.

pingskills commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -117,7 +117,7 @@ en:
         pending_avatar: "After uploading a new avatar, you will have to wait for the Board to approve it."
       competition:
         competition_id_to_clone: "Optional. Can save you some typing if you're running a similar competition to a past competition."
-        name: "The full name of the competition including the year at the end (e.g., European Rubik's Cube Championship 2006). Be sure to capitalize."
```
The example name was longer than the suggested 32 characters to I changed it to a competition name that passes the warnings.

pingskills commented:
OK. So I need to also commit the structure.sql file and Travis uses this to determine what migrations have run?

jfly commented:
Rails looks at the schema_migrations table to know what migrations have been run. Unfortunately, when you run a migration locally, a lot of unrelated pieces of that structure file change. To avoid problems, can you remove all the unrelated parts of the diff first?

pingskills commented:
Hi @jfly,
if I understand correctly then I should change the line in my file from:
`Server version	10.1.14-MariaDB`
to
`Server version	5.7.12-0ubuntu1`

And then update all similar unrelated changes so that only the relevant differences are shown and the add this **manually** modified file to the commit. Is that the right process?

jfly commented:
That would work, but would be kind of painful. The tool I use I `git add -p` to just say yes and no to all the pieces of the file that changed.

pingskills commented:
Cool. Thanks for the tip @jfly. I will checkout `git add -p`

pingskills commented:
Fixed up structure.sql - Let me know your thoughts on the PR.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Is this a rubocop thing?

pingskills replied:
Yes @jonatanklosko. First I had to add .freeze to a definition and then I had to add the `frozen_string_literal: true`

jonatanklosko replied:
Ok, so I think it wouls be better to do the configuration  in .rubocop.yml

jfly replied:
@jonatanklosko, this actually isn't a rubocop comment. This is a comment that tells Ruby 2.3 to treat all string allocations in this file as immutable strings. @pingskills, I believe the fact that you added this comment means you can actually remove the call to `.freeze` on line 52.

The rubocop @pingskills ran into appears to be rubocop helping force us to prepare for Ruby 3.0 where strings will be immutable by default. Here's what I see if I run rubocop on @pingskills's project after removing this comment:

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bundle exec rubocop
Inspecting 279 files
.................................................................................................................................................C.....................................................................................................................................

Offenses:

app/models/competition.rb:1:1: C: Missing frozen string literal comment.
class Competition < ActiveRecord::Base
^

279 files inspected, 1 offense detected
```

I support this comment. In a later PR, I'd like to turn this on everywhere. If we make use of immutable strings anywhere, I'm going to be rather surprised.

pingskills replied:
I just tried removing the .freeze on line 52 but rubocop wasn't happy :)

`
Offenses:
app/models/competition.rb:53:26: C: Freeze mutable objects assigned to constants.
  INVALID_NAME_MESSAGE = "must end with a year and must contain only alphnumeric characters, dashes(-), ampersands(&), periods(.), colons(:), apostrophes('), and spaces( )"
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^`

larspetrus commented:
It's a Ruby 2.3 feature: https://wyeworks.com/blog/2015/12/1/immutable-strings-in-ruby-2-dot-3

It probably breaks things if done project wide.

pingskills commented:
Deployed.

jfly commented:
Great. Thanks Jeff!
